File: src/main/java/com/dems/cases/CaseService.java
package com.dems.cases;

import com.dems.enums.CaseStatus;
import com.dems.enums.CrimeSeverity;
import com.dems.enums.CrimeType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;

/**
 * Service interface managing Case lifecycle, status transitions, officer assignments, and dynamic searching.
 */
public interface CaseService {

    CaseResponse createCase(CaseCreateRequest request);

    CaseResponse getCaseById(Long id);

    Page<CaseSummaryResponse> getAllCases(Pageable pageable);

    Page<CaseSummaryResponse> getMyAssignedCases(String userEmail, Pageable pageable);

    CaseResponse updateCase(Long id, CaseUpdateRequest request);

    CaseResponse updateCaseStatus(Long id, CaseStatusUpdateRequest request);

    CaseResponse assignOfficer(Long id, CaseAssignOfficerRequest request);

    Page<CaseSummaryResponse> searchCases(
            String caseNumber,
            String crimeNumber,
            String caseName,
            CaseStatus status,
            CrimeType crimeType,
            CrimeSeverity severity,
            Long assignedOfficerId,
            LocalDate incidentDate,
            Boolean active,
            String searchTerm,
            Pageable pageable
    );
}

File: frontend/src/services/caseService.js
import apiClient from './apiClient';
import { API_ENDPOINTS } from '../constants/apiEndpoints';

export const caseService = {
  createCase: async (requestData) => {
    return await apiClient.post(API_ENDPOINTS.CASES.BASE, requestData);
  },

  getAllCases: async (params = {}) => {
    return await apiClient.get(API_ENDPOINTS.CASES.BASE, { params });
  },

  getMyAssignedCases: async (params = {}) => {
    return await apiClient.get(API_ENDPOINTS.CASES.MY_CASES, { params });
  },

  getCaseById: async (id) => {
    return await apiClient.get(API_ENDPOINTS.CASES.BY_ID(id));
  },

  updateCase: async (id, requestData) => {
    return await apiClient.put(API_ENDPOINTS.CASES.BY_ID(id), requestData);
  },

  updateCaseStatus: async (id, statusData) => {
    return await apiClient.patch(API_ENDPOINTS.CASES.STATUS(id), statusData);
  },

  assignOfficer: async (id, assignData) => {
    return await apiClient.patch(API_ENDPOINTS.CASES.ASSIGN_OFFICER(id), assignData);
  },

  searchCases: async (searchParams = {}) => {
    return await apiClient.get(API_ENDPOINTS.CASES.SEARCH, { params: searchParams });
  },
};

export default caseService;
